mirascope/tools/base.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import annotations
+
+import inspect
+from abc import ABC
+from typing import ClassVar, Generic, TypeVar, cast
+
+from pydantic import BaseModel, ConfigDict, create_model
+
+from mirascope.core import BaseTool
+
+
+class _ToolConfig(BaseModel, ABC):
+    """Base configuration for tools"""
+
+    @classmethod
+    def from_env(cls: type[_ToolConfigT]) -> _ToolConfigT:
+        """Load configuration from environment variables.
+        By default, returns an instance with default values."""
+        return cls()
+
+
+_ToolConfigT = TypeVar("_ToolConfigT", bound=_ToolConfig)
+
+_ToolSchemaT = TypeVar("_ToolSchemaT")
+
+
+class ConfigurableTool(
+    BaseTool[_ToolSchemaT], Generic[_ToolConfigT, _ToolSchemaT], ABC
+):
+    """
+    Abstract base class for configurable tools.
+
+    Subclasses must define a `__prompt_usage_description__` class variable
+    and __config__ class variable with a subclass of _ToolConfig.
+    """
+
+    __config__: ClassVar[_ToolConfig]
+    __prompt_usage_description__: ClassVar[str]
```

Comment:
I think we need to decide on some form of "standard" for the prompt usage descriptions so that all pre-made tools can be used the same way. Right now I'm noticing that the usage descriptions don't necessarily follow a standard structure or format.

What do you think about something like this:

```
"""
- `{TOOL_NAME}`: {short_description}
   - {bullets describing use} 
"""
```

For example, I imagine the ParseURLContent usage description would look something like this:

```
"""
- `ParseURLContent`: Returns the given URL's main content as clean text
    - Fetches the raw HTML content for the given URL
    - Removes all unwanted elements (scripts, styles, navigation, etc.)
    - Attempts to find the main content section
    - If no main content section is found, extracts all text content
    - Returns cleaned and formatted text content
    - If any errors occur during fetching or parsing, and error message is returned.
"""
```

WDYT?